### PR TITLE
feat(ui): linkify https URLs in the admin site banner

### PR DESCRIPTION
**Pull request by ishiland**
The admin announcement banner (SiteBanner.tsx) rendered banner_text as plain escaped text, so a URL in it displayed but couldn't be clicked. Matched https:// URLs now render as anchors (target=_blank, rel=noopener noreferrer, trailing sentence punctuation kept outside the link); all other text remains React-escaped, so the banner still can't carry injected markup. No new setting, no API/schema change.

Verification: SiteBanner suite 7/7 (2 new cases: linkify + escaping/no-http), eslint clean, typecheck clean. Counterfactual: with the component change reverted and tests kept, exactly the 2 new tests fail.

### CODE REVIEW

**Comment by ishiland** [comment]
@codex review

**Comment by ishiland** [comment]
@codex review

File: frontend/src/components/layout/SiteBanner.tsx [hunk 1]
```diff
@@ -29,6 +29,40 @@ const COLOR_CLASSES: Record<string, string> = {
 
 const DISMISS_KEY = 'gl-site-banner-dismissed';
 
+// feat(#123-followup): make https:// URLs in admin-entered banner text
```
[the remark this answers is not in the document]
Comment (ishiland) [reply frontend/src/components/layout/SiteBanner.tsx]: Fixed in 741804dd3 — the comment now anchors feat(#1662).

File: frontend/src/components/layout/SiteBanner.tsx [hunk 2]
```diff
@@ -29,6 +29,40 @@ const COLOR_CLASSES: Record<string, string> = {
...
+// clickable. The text itself stays React-escaped; only the matched URL
+// becomes an anchor, so an admin cannot inject markup through the banner.
+// Trailing sentence punctuation stays outside the link.
+const URL_RE = /(https:\/\/[^\s<>"']+?)([.,;:!?)]*)(?=\s|$)/g;
```
[the remark this answers is not in the document]
Comment (ishiland) [reply frontend/src/components/layout/SiteBanner.tsx]: Fixed in 741804dd3 — a paren-balance pass keeps as many closing parens in the URL as it has unmatched openers (test H covers the Wikipedia case and the parenthesized-URL case).

File: frontend/src/components/layout/SiteBanner.tsx [hunk 3]
```diff
@@ -29,6 +29,56 @@ const COLOR_CLASSES: Record<string, string> = {
 
 const DISMISS_KEY = 'gl-site-banner-dismissed';
 
+// feat(#1662): make https:// URLs in admin-entered banner text
+// clickable. The text itself stays React-escaped; only the matched URL
+// becomes an anchor, so an admin cannot inject markup through the banner.
+// Trailing sentence punctuation stays outside the link.
+const URL_RE = /(https:\/\/[^\s<>"']+?)([.,;:!?)]*)(?=\s|$)/g;
```
[the remark this answers is not in the document]
Comment (ishiland) [reply frontend/src/components/layout/SiteBanner.tsx]: Fixed in 77f8c5ed6 — the lookahead now accepts the excluded delimiters as boundaries, so quoted and angle-bracketed URLs linkify (test I).

**Comment by ishiland** [comment]
@codex review

File: frontend/src/components/layout/SiteBanner.tsx [hunk 4]
```diff
@@ -29,6 +29,56 @@ const COLOR_CLASSES: Record<string, string> = {
 
 const DISMISS_KEY = 'gl-site-banner-dismissed';
 
+// feat(#1662): make https:// URLs in admin-entered banner text
+// clickable. The text itself stays React-escaped; only the matched URL
+// becomes an anchor, so an admin cannot inject markup through the banner.
+// Trailing sentence punctuation stays outside the link.
+const URL_RE = /(https:\/\/[^\s<>"']+?)([.,;:!?)]*)(?=[\s<>"']|$)/g;
```
[the remark this answers is not in the document]
Comment (ishiland) [reply frontend/src/components/layout/SiteBanner.tsx]: Fixed in 7d754c157 — the trailing class now includes ']' and the balance pass generalizes over both bracket kinds: wrapped URLs shed the wrapper, balanced in-path brackets survive (test J covers [wrapped] and /item[3]).

**Comment by ishiland** [comment]
@codex review

File: frontend/src/components/layout/SiteBanner.tsx [hunk 5]
```diff
@@ -29,6 +29,62 @@ const COLOR_CLASSES: Record<string, string> = {
 
 const DISMISS_KEY = 'gl-site-banner-dismissed';
 
+// feat(#1662): make https:// URLs in admin-entered banner text
+// clickable. The text itself stays React-escaped; only the matched URL
+// becomes an anchor, so an admin cannot inject markup through the banner.
+// Trailing sentence punctuation stays outside the link.
+const URL_RE = /(https:\/\/[^\s<>"']+?)([.,;:!?)\]]*)(?=[\s<>"']|$)/g;
```
[the remark this answers is not in the document]
Comment (ishiland) [reply frontend/src/components/layout/SiteBanner.tsx]: Fixed in d50d5867a — curly quotes and guillemets joined the excluded/boundary classes, and …、。 joined the trailing class; test K pins the smart-quoted and ellipsis-suffixed cases.

**Comment by ishiland** [comment]
@codex review

File: frontend/src/components/layout/SiteBanner.tsx [hunk 6]
```diff
@@ -29,6 +29,63 @@ const COLOR_CLASSES: Record<string, string> = {
 
 const DISMISS_KEY = 'gl-site-banner-dismissed';
 
+// feat(#1662): make https:// URLs in admin-entered banner text
+// clickable. The text itself stays React-escaped; only the matched URL
+// becomes an anchor, so an admin cannot inject markup through the banner.
+// Trailing sentence punctuation stays outside the link.
+const URL_RE =
+  /(https:\/\/[^\s<>"'\u201C\u201D\u2018\u2019\u00AB\u00BB]+?)([.,;:!?)\]\u2026\u3002\u3001]*)(?=[\s<>"'\u201C\u201D\u2018\u2019\u00AB\u00BB]|$)/g;
```
[the remark this answers is not in the document]
Comment (ishiland) [reply frontend/src/components/layout/SiteBanner.tsx]: Fixed in e9c519d49 — apostrophe left the excluded class (in-URL apostrophes survive: test L pins /O'Reilly), and a trailing apostrophe is shed only when paired with an opening quote immediately before the URL ('wrapped' case).

**Comment by ishiland** [comment]
@codex review

File: frontend/src/components/layout/SiteBanner.tsx [hunk 7]
```diff
@@ -29,6 +29,73 @@ const COLOR_CLASSES: Record<string, string> = {
...
 
+// feat(#1662): make https:// URLs in admin-entered banner text
+// clickable. The text itself stays React-escaped; only the matched URL
+// becomes an anchor, so an admin cannot inject markup through the banner.
+// Trailing sentence punctuation stays outside the link.
+// Apostrophes are legitimate URL characters (/O'Reilly), so ' is NOT an
+// excluded body character; a single quote only acts as a delimiter when it
+// wraps the URL ('https://…'), handled by the paired-wrapper check below.
+const URL_RE =
+  /(https:\/\/[^\s<>"\u201C\u201D\u2018\u2019\u00AB\u00BB]+?)([.,;:!?)\]\u2026\u3002\u3001]*)(?=[\s<>"\u201C\u201D\u2018\u2019\u00AB\u00BB]|$)/g;
```
[the remark this answers is not in the document]
Comment (ishiland) [reply frontend/src/components/layout/SiteBanner.tsx]: Fixed in a9b9963ab — the regex carries the i flag; URI schemes are case-insensitive and the href keeps the admin's casing (test M).

[the remark this answers is not in the document]
Comment (ishiland) [reply frontend/src/components/layout/SiteBanner.tsx]: Fixed in a9b9963ab — <URL> is RFC 3986's own delimiting, so an angle-bracketed URL is now taken VERBATIM (no trail or wrapper heuristics; test M pins /wiki/Yahoo!). That also makes <…> the documented escape hatch for any exact URL the heuristics would trim — which is where the delimiter model closes: further undelimited edge cases have that escape available.

**Comment by ishiland** [comment]
@codex review

File: frontend/src/components/layout/SiteBanner.tsx [hunk 8]
```diff
@@ -29,6 +29,85 @@ const COLOR_CLASSES: Record<string, string> = {
...
 
+// feat(#1662): make https:// URLs in admin-entered banner text
+// clickable. The text itself stays React-escaped; only the matched URL
+// becomes an anchor, so an admin cannot inject markup through the banner.
+// Trailing sentence punctuation stays outside the link.
+// Apostrophes are legitimate URL characters (/O'Reilly), so ' is NOT an
+// excluded body character; a single quote only acts as a delimiter when it
+// wraps the URL ('https://…'), handled by the paired-wrapper check below.
+const URL_RE =
+  /(https:\/\/[^\s<>"\u201C\u201D\u2018\u2019\u00AB\u00BB]+?)([.,;:!?)\]\u2026\u3002\u3001]*)(?=[\s<>"\u201C\u201D\u2018\u2019\u00AB\u00BB]|$)/gi;
```
[the remark this answers is not in the document]
Comment (ishiland) [reply frontend/src/components/layout/SiteBanner.tsx]: Fixed in 419b9c59f — the lazy-body/trail-group split is gone: one greedy match with no lookahead, trail peeled by an end-anchored regex in code. Linear by construction; all 13 behavior tests unchanged and green.

**Comment by ishiland** [comment]
@codex review

File: frontend/src/components/layout/SiteBanner.tsx [hunk 9]
```diff
@@ -29,6 +29,91 @@ const COLOR_CLASSES: Record<string, string> = {
...
+// becomes an anchor, so an admin cannot inject markup through the banner.
+// Trailing sentence punctuation stays outside the link.
+// Apostrophes are legitimate URL characters (/O'Reilly), so ' is NOT an
+// excluded body character; a single quote only acts as a delimiter when it
+// wraps the URL ('https://…'), handled by the paired-wrapper check below.
+// fix(#1662 review): a single greedy group with no trailing-group/lookahead
+// split — the earlier lazy-body + punctuation-group pair let the engine retry
+// every split point on adversarial runs (quadratic). The trail is peeled in
+// plain code below instead, which is linear by construction.
+const URL_RE = /https:\/\/[^\s<>"\u201C\u201D\u2018\u2019\u00AB\u00BB]+/gi;
```
[the remark this answers is not in the document]
Comment (ishiland) [reply frontend/src/components/layout/SiteBanner.tsx]: Fixed in b4546949a — a left-boundary check skips matches whose preceding character isn't a delimiter, so nothttps:// and URL-embedded https:// stay plain (test N).

**Comment by ishiland** [comment]
@codex review

File: frontend/src/components/layout/SiteBanner.tsx [hunk 10]
```diff
@@ -29,6 +29,97 @@ const COLOR_CLASSES: Record<string, string> = {
...
```
[the remark this answers is not in the document]
Comment (ishiland) [reply frontend/src/components/layout/SiteBanner.tsx]: Fixed in 65525f18b — em/en dashes joined the excluded class (they end an undelimited URL: test O), and the <URL> verbatim branch now spans to the closing bracket itself so a dash inside <…> survives. With that, the delimiter model is closed: <URL> is the escape hatch for anything the heuristics would trim.

**Comment by ishiland** [comment]
@codex review

File: frontend/src/components/layout/SiteBanner.tsx [hunk 11]
```diff
@@ -29,6 +29,101 @@ const COLOR_CLASSES: Record<string, string> = {
...
+const TRAIL_RE = /[.,;:!?)\]\u2026\u3002\u3001]+$/;
+
+function renderBannerText(text: string) {
+  const parts: Array<string | { url: string; trail: string }> = [];
+  let last = 0;
+  for (const m of text.matchAll(URL_RE)) {
+    // fix(#1662 review): only linkify at a token boundary — an https:// embedded
+    // inside another token (nothttps://…, or a URL nested in another URL's
+    // path) is not a link of its own.
+    if (m.index! > 0 && !/[\s<>"'()[\]\u201C\u201D\u2018\u2019\u00AB\u00BB]/.test(text[m.index! - 1])) {
```
[the remark this answers is not in the document]
Comment (ishiland) [reply frontend/src/components/layout/SiteBanner.tsx]: Fixed in cda9db7dd — the dashes joined the left-boundary class (symmetry with their right-delimiter role; test P). Same push rebases onto main for the #1663 changelog landing.

**Comment by ishiland** [comment]
@codex review

File: frontend/src/components/layout/SiteBanner.tsx [hunk 12]
```diff
@@ -29,6 +29,101 @@ const COLOR_CLASSES: Record<string, string> = {
...
+// split — the earlier lazy-body + punctuation-group pair let the engine retry
+// every split point on adversarial runs (quadratic). The trail is peeled in
+// plain code below instead, which is linear by construction.
+const URL_RE = /https:\/\/[^\s<>"\u201C\u201D\u2018\u2019\u00AB\u00BB\u2013\u2014]+/gi;
+const TRAIL_RE = /[.,;:!?)\]\u2026\u3002\u3001]+$/;
+
+function renderBannerText(text: string) {
+  const parts: Array<string | { url: string; trail: string }> = [];
+  let last = 0;
+  for (const m of text.matchAll(URL_RE)) {
```
[the remark this answers is not in the document]
Comment (ishiland) [reply frontend/src/components/layout/SiteBanner.tsx:57]: Fixed in ad501fc4d — matches starting before `last` are skipped, so a scheme inside an already-consumed <…> span can't emit an overlapping link (test Q). Good catch — this one was a real logic bug, not a delimiter edge.

**Comment by ishiland** [comment]
@codex review

File: frontend/src/components/layout/SiteBanner.tsx [hunk 13]
```diff
@@ -29,6 +29,105 @@ const COLOR_CLASSES: Record<string, string> = {
...
+    let trail = trailMatch ? trailMatch[0] : '';
+    // fix(#1662 review): <https://…> is RFC 3986's own URL delimiting. Inside
+    // angle brackets the URL is taken VERBATIM — trailing punctuation like
+    // /wiki/Yahoo! stays in the href, and no wrapper/balance heuristics run.
+    // This is also the documented escape hatch: an admin who needs an exact
+    // URL that the heuristics would trim can always write <URL>.
+    // The regex may stop early inside <…> (excluded characters like dashes are
+    // allowed there), so the verbatim span runs to the closing bracket itself.
+    if (m.index! > 0 && text[m.index! - 1] === '<') {
+      const close = text.indexOf('>', m.index!);
```
[the remark this answers is not in the document]
Comment (ishiland) [reply frontend/src/components/layout/SiteBanner.tsx]: Fixed in ccc77d2bd — the '>' search is bounded to the current whitespace-free token (slice + indexOf), so repeated unclosed '<' spans stay linear; an unclosed '<' falls back to normal linkify (test R).

**Comment by ishiland** [comment]
@codex review

File: frontend/src/components/layout/SiteBanner.tsx [hunk 14]
```diff
@@ -29,6 +29,110 @@ const COLOR_CLASSES: Record<string, string> = {
...
+    // URL that the heuristics would trim can always write <URL>.
+    // The regex may stop early inside <…> (excluded characters like dashes are
+    // allowed there), so the verbatim span runs to the closing bracket itself.
+    if (m.index! > 0 && text[m.index! - 1] === '<') {
+      // fix(#1662 review): bound the '>' search to the current whitespace-free
+      // token — an unclosed '<' must not scan the whole remaining text, or
+      // repeated unclosed spans would make the render quadratic.
+      const ws = text.slice(m.index!).search(/\s/);
+      const tokenEnd = ws === -1 ? text.length : m.index! + ws;
+      const closeInToken = text.slice(m.index!, tokenEnd).indexOf('>');
```
[the remark this answers is not in the document]
Comment (ishiland) [reply frontend/src/components/layout/SiteBanner.tsx:86]: Fixed in 2b38cb17d — a 2,000-char linkification cap bounds worst-case cost to a constant (longer banners render as plain text; test S). This structurally closes the adversarial-complexity class rather than proving each scan linear.

**Comment by ishiland** [comment]
@codex review

File: frontend/src/components/layout/SiteBanner.tsx [hunk 15]
```diff
@@ -29,6 +29,118 @@ const COLOR_CLASSES: Record<string, string> = {
...
+        continue;
+      }
+    }
+    // fix(#1662 review): 'https://…' wrapped in single quotes — the opening
+    // quote sits just before the match, so a trailing apostrophe is the
+    // closing wrapper, not part of the URL. Unpaired apostrophes stay in.
+    if (m.index! > 0 && text[m.index! - 1] === "'" && url.endsWith("'")) {
+      url = url.slice(0, -1);
+      trail = "'" + trail;
+    }
```
[the remark this answers is not in the document]
Comment (ishiland) [reply frontend/src/components/layout/SiteBanner.tsx:100]: Declining this one. The failing shape is an admin who single-quote-wraps a URL AND puts sentence punctuation inside the quotes — banner text is admin-entered, previewed the moment it's saved, and the <URL> form (or writing the URL bare) is the documented exact-URL path. After 13 addressed findings the heuristics are at equilibrium: this round's ask (peel more inside quote wrappers) and round 8's ask (stricter token boundaries) now trade against each other, which is where a prose linkifier for a one-line admin banner should stop.

File: frontend/src/components/layout/SiteBanner.tsx [hunk 16]
```diff
@@ -29,6 +29,118 @@ const COLOR_CLASSES: Record<string, string> = {
...
+  let last = 0;
+  for (const m of text.matchAll(URL_RE)) {
+    // fix(#1662 review): a verbatim <…> span advances `last` past characters
+    // matchAll has not seen; a scheme inside the consumed span must not
+    // produce a second, overlapping link.
+    if (m.index! < last) continue;
+    // fix(#1662 review): only linkify at a token boundary — an https:// embedded
+    // inside another token (nothttps://…, or a URL nested in another URL's
+    // path) is not a link of its own.
+    if (m.index! > 0 && !/[\s<>"'()[\]\u201C\u201D\u2018\u2019\u00AB\u00BB\u2013\u2014]/.test(text[m.index! - 1])) {
```
[the remark this answers is not in the document]
Comment (ishiland) [reply frontend/src/components/layout/SiteBanner.tsx:65]: Declining this one for the same reason as its sibling: 'Details:https://…' (no space) is an admin typo with immediate visual feedback, and loosening the left-boundary class works against the round-8 requirement that embedded schemes (nothttps://, URL-in-URL) stay plain. The admin fix is a space; the exact-URL path is <URL>. Boundary model closed.
